scripts/west_commands/run_common.py

```diff
@@ -145,21 +147,49 @@ def add_parser_common(command, parser_adder=None, parser=None):
 
     return parser
 
-def do_run_common(command, user_args, user_runner_args):
+def do_run_common(command, user_args, user_runner_args, multi_domain=False):
```

Comment:
I'm missing the high level architectural view of how this is meant to work. I read the commit log and re-read the architecture slides which show that the build system generates a domain runners file, but I'm missing some architecture level understanding of how this is meant to affect the runners. This is introducing a notion of 'single-domain' versus 'multi-domain' runner invocation, and I'm not sure I follow all the details from a first quick skim.

I guess we can discuss further offline if you want to have a high bandwidth conversation about it, otherwise I'll try to check the PR again soon.

Reply:
Hopefully the docs i'm writing for SYSBuild will help to understand the need, but for now, let me give a quick reasoning.

SYSBuild creates `domain_runners.yaml`.
The runners for each domain are executed in serial, unless user specify `--domain mcuboot` in which case only mcuboot is flashed.
This means per default flashing is done for each domain in sequence, first MCUboot, thereafter the `app`.

However, for debugging, it's impractical to launch all debug sessions in sequence.
For debugging, only a single domain runner should be invoked.
So if no domain is given, then the `app` will be default, but user can still decide to debug MCUboot with `west debug --domain mcuboot`.

So the `multi_domain` argument here simply tells if a given runner default should be invoked on all domains (flash), or just a single (debug).

See also: https://github.com/zephyrproject-rtos/zephyr/pull/40555#issuecomment-1011874931

Reply:
This seems to be arguing against a `do_run_common()`, though. It feels like flashing and debugging are separate enough tasks in a multi-domain world that I'm not sure it makes sense to keep common code for both of them. I need to think about this some more, though. I appreciate the short explanation but I would feel more comfortable with a breakdown of use cases and how they're being addressed here.

Reply:
@mbolivar-nordic I have reworked the domain handling in the west extensions.
A new `domains.py` has also been introduced.

docs. especially `west build|flash|debug` sections, has been improved with domain description.
Hopefully all of this should make the design and user interface much clearer.